### DISCUSSION TITLE
Automatic management of ESP32 flash partitions

### DISCUSSION THREAD

**Discussion by phoddie** [Moddable-OpenSource/moddable#1087] [Announcements]
There's a very cool new feature in the `mcconfig` tool that automates management of flash [partitions on ESP32](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/partition-tables.html) builds. Partitions are how the ESP-IDF divides flash storage for different uses including  your project code, file system storage, NVS (used by Moddable SDK Preferences), and [mods](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/mods.md). There are also special partition requirements to support Over-the-Air (OTA) firmware updates.

Before these changes, developers had to manage partitions manually by creating a partitions.csv file for their project to override the defaults. The defaults reserved space for features, even if a particular project didn't use them. So, if your project didn't use a file system, for example, space was still reserved for files, reducing the flash space available for your project. Even a simple partition map is a bit complicated to understand, which discourages developers from modifying them. For example, here's our default ESP32 partition map before these changes:

```
nvs,      data, nvs,     0x9000,  0x006000,
phy_init, data, phy,     0xf000,  0x001000,
factory,  app,  factory, 0x10000, 0x390000,
xs,       0x40, 1,       0x3A0000, 0x040000,
settings, data, 1,       0x3E0000, 0x010000,
storage,  data, spiffs,  0x3F0000, 0x010000,
```

With the latest changes, `mcconfig` takes care of generating a custom partitions.csv automatically. You can still partition flash yourself, for special project requirements. But, for most developers, these changes will provide more flash space for their projects and eliminate the need to update the partition map themselves. It also simplifies porting to new ESP32 boards. Here's the ESP32 partition map after the changes:

```
nvs,      data, nvs,     0x9000,  0x006000,
phy_init, data, phy,     0xf000,  0x001000,
factory,  app,  factory, 0x10000, 0x3F0000,
```

The way `mcconfig` does this is by first looking for a "factory partition." For projects that don't use OTA, the factory partition holds the firmware image. If there is a factory partition, `mcconfig` can performa automatic partitioning. Modules like OTA and Files signal in their manifests that certain features are being used using the defines feature of manifests. Then `mcconfig` uses that information to divide up the factory partition as needed:

- If mod support is enabled, a partition for mods is created.
- If a file system is used ([LittleFS](https://blog.moddable.com/blog/littlefs/), SPIFFS, or FAT32), a file system partition is created.
- If OTA updates are used, 8 KB is reserved for the "OTA Data" partition required by the ESP-IDF. The remaining space is divided into two parts to create the OTA partitions.

There are other rules, like OTA partitions must begin on 64 KB boundaries, that `mcconfig` automatically applies.

This feature is entirely transparent. By simply adding modules to your project, the partition map is automatically updated for you. The partition map now always make optimal use of the available space.

To learn more about how automatic partition for ESP32 works, check out the [updated manifest documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/manifest.md#how-partitionscsv-is-processed).

**Comment by phoddie** [comment]
I noticed that M5Stack Core2 port is only 4 MB of the 16 MB flash. That's because it is using the default partition map for ESP32 ports, which is 4 MB. I added a simple [partitions.csv](https://github.com/Moddable-OpenSource/moddable/blob/public/build/devices/esp32/targets/m5stack_core2/sdkconfig/partitions.csv) and [sdkconfig.defaults](https://github.com/Moddable-OpenSource/moddable/blob/public/build/devices/esp32/targets/m5stack_core2/sdkconfig/sdkconfig.defaults) and included them in the [manifest](https://github.com/Moddable-OpenSource/moddable/blob/bdc39c559875a75c1a9b3c051457d6952c114d99/build/devices/esp32/targets/m5stack_core2/manifest.json#L5-L6). Now all 16 MB is available!

The extra space should be nice for [Stack-chan](https://github.com/meganetaaan/stack-chan). These changes should also be a good model for other ports to follow for ESP32 devices with larger flash memory.

**Comment by AlexLucardGoodman** [comment]
Good afternoon. I'm back to trying some stuff with the esp32 and moddable SDK. Id like to get the OTA updates working. I activated the OTA autosplit in manifest.json, but I cant find much info or documentation about a moddable sdk OTA module. Could you point me in the right direction, please? Thank you!

**Reply by phoddie** [reply]
Hello.

There are two generations of the OTA module. There is the original for ESP32 and then new ECMA-419 standard Update module. I strongly recommend using the Update module, as the other isn't getting attention these days and will eventually be removed.

The Update module is new in ECMA-419 3rd Edition. The API has a small surface area, so there isn't too much to learn. There are resources available to help:

- The [standard text](https://419.ecma-international.org/#-26-persistent-storage-update) itself which thoroughly describes the API
- The [unit tests](https://github.com/Moddable-OpenSource/moddable/tree/public/tests/modules/io/update) use to verify implementations. 
- Implementations for [ESP32](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/io/update/esp32) and the [macOS simulator](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/io/update/mac)
